### PR TITLE
Support signing a pre-calculated hash

### PR DESCRIPTION
**Pull request by jls5177**
This code change adds a new function that takes a pre-computed hash of a message that should be signed (as apposed to a message). This change allows one process to generate the hash of message and another process (over RPC in my case) to perform the signing.

### CODE REVIEW

File: rsa/pkcs1.py [hunk 1]
```diff
@@ -283,6 +279,32 @@ def sign(message, priv_key, hash):
...
+    :param hash_method: the hash method used on the message. Use 'MD5', 'SHA-1',
+        'SHA-256', 'SHA-384' or 'SHA-512'.
+    :return: a message signature block.
+    :raise OverflowError: if the private key is too small to contain the
+        requested hash.
+
+    """
+    
+    # Verify hash_method is a valid hash algorithm
+    if hash_method not in HASH_ASN1:
```
Comment (sybrenstuvel) [inline rsa/pkcs1.py]: Why perform this check here? It's the same as in `sign_hash()` anyway.

Comment (jls5177) [inline rsa/pkcs1.py]: This was an oversight in my original commit. This is removed in the new commit.

**Comment by jls5177** [comment]
I added 2 new unit tests to cover the new hashing flow and updated the usage docs with the new flow. I am new to this type of documentation. Let me know if I missed anything.

File: rsa/pkcs1.py [hunk 2]
```diff
@@ -283,6 +279,28 @@ def sign(message, priv_key, hash):
...
+    :param hash_method: the hash method used on the message. Use 'MD5', 'SHA-1',
+        'SHA-256', 'SHA-384' or 'SHA-512'.
+    :return: a message signature block.
+    :raise OverflowError: if the private key is too small to contain the
+        requested hash.
+
+    """
+
+    # Calculate the hash and perform the signing
+    return sign_hash(hash(message, hash_method), priv_key, hash_method)
```
Comment (sybrenstuvel) [inline rsa/pkcs1.py]: Please splitt his up into two lines, using a helper variable. The way it is now it does too much on one line; a traceback referring to an error in this line will be too ambiguous.

File: tests/test_pkcs1.py [hunk 3]
```diff
@@ -106,3 +106,35 @@ def test_multiple_signings(self):
...
+        message = b'je moeder'
+        print("\tMessage:   %r" % message)
+
+        msg_hash = pkcs1.hash(message, 'SHA-256')
+        print("\tHash: %r" % msg_hash)
+
+        signature1 = pkcs1.sign_hash(msg_hash, self.priv, 'SHA-256')
+        print("\tSignature1: %r" % signature1)
+
+        # Calculate the signature using the original method
```
Comment (sybrenstuvel) [inline tests/test_pkcs1.py]: The comment "the original method" implies that the reader knows that your method is something that was added later. This is only true within the context of this pull request, but not later once everything has been merged.

File: tests/test_pkcs1.py [hunk 4]
```diff
@@ -106,3 +106,35 @@ def test_multiple_signings(self):
         signature2 = pkcs1.sign(message, self.priv, 'SHA-1')
 
         self.assertEqual(signature1, signature2)
+
+    def test_split_hash_sign(self):
+        """Hashing and then signing should match with directly signing the message. """
+
+        message = b'je moeder'
+        print("\tMessage:   %r" % message)
```
Comment (sybrenstuvel) [inline tests/test_pkcs1.py]: Please don't print things in a unit test.

Comment (jls5177) [inline tests/test_pkcs1.py]: I was using the other test cases in this file as a template and many of them had prints.  The prints are now removed.

File: tests/test_pkcs1.py (hunk 3 above)
Comment (jls5177) [inline tests/test_pkcs1.py]: Agreed, I switched it to now say `unified method`.

File: rsa/pkcs1.py (hunk 2 above)
Comment (jls5177) [inline rsa/pkcs1.py]: Sure, this will be fixed in the next commit.

**Comment by sybrenstuvel** [comment]
Another thing: `hash()` is not a good name for a function, as it's also [a built-in Python function](https://docs.python.org/3/library/functions.html#hash).

**Comment by sybrenstuvel** [comment]
Hmm not too fond of "generate", maybe "compute_hash"?

**Comment by sybrenstuvel** [comment]
Thanks, it's in!

**Pull request merged**